### PR TITLE
[FLINK-6073] [table] Support for SQL inner queries for proctime

### PR DESCRIPTION
**Pull request by rtudoran**
Thanks for contributing to Apache Flink. Before you open your pull request, please take the following check list into consideration.
If your changes take all of the items into account, feel free to open your pull request. For more information and/or questions please refer to the [How To Contribute guide](http://flink.apache.org/how-to-contribute.html).
In addition to going through the list, please provide a meaningful description of your changes.

- [x ] General
  - The pull request references the related JIRA issue ("[FLINK-XXX] Jira title text")
  - The pull request addresses only one issue
  - Each commit in the PR has a meaningful commit message (including the JIRA id)

- [x ] Documentation
  - Documentation has been added for new functionality
  - Old documentation affected by the pull request has been updated
  - JavaDoc for public methods has been added

- [ ] Tests & Build
  - Functionality added by the pull request is covered by tests
  - `mvn clean verify` has been executed successfully locally or a Travis build has passed

### CODE REVIEW

**Comment by rtudoran** [comment]
@fhueske @twalthr @sunjincheng121 @shijinkui @stefanobortoli @hongyuhong

I have made a first implementation draft for supporting inner queries mainly when operating on processing time. I would highly appreciate some feedback from you to further enhance the approach.

The idea of the implementation is described in https://issues.apache.org/jira/browse/FLINK-6073?filter=-2

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/table/plan/nodes/datastream/DataStreamJoin.scala [hunk 1]
```diff
@@ -0,0 +1,241 @@
...
+import java.lang.Iterable
+import org.apache.flink.streaming.runtime.streamrecord.StreamRecord
+import org.apache.flink.streaming.runtime.operators.windowing.TimestampedValue
+import org.apache.flink.api.common.functions.RichFlatJoinFunction
+import org.apache.flink.configuration.Configuration
+import org.apache.flink.api.common.state.ValueState
+import org.apache.flink.api.common.state.ValueStateDescriptor
+import org.apache.flink.util.Collector
+
+class DataStreamJoin(
```
Comment (shijinkui) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/table/plan/nodes/datastream/DataStreamJoin.scala]: need scaladoc to describe the class's responsibility

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/table/plan/nodes/datastream/DataStreamJoin.scala [hunk 2]
```diff
@@ -0,0 +1,241 @@
...
+    val rowTypeInfo = FlinkTypeFactory.toInternalRowTypeInfo(getRowType).asInstanceOf[RowTypeInfo]
+
+    val result = inputDSLeft.join(inputDSRight)
+      .where(new EmptyKeySelector()).equalTo(new EmptyKeySelector())
+      .window(GlobalWindows.create())
+      .trigger(new ProcTimeLeftJoinTrigger())
+      .evictor(new FullEvictor())
+      .apply(new JoinProcTimeForInnerQuerry(rowTypeInfo))
+
+    null
```
Comment (shijinkui) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/table/plan/nodes/datastream/DataStreamJoin.scala]: why return null

Comment (rtudoran) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/table/plan/nodes/datastream/DataStreamJoin.scala]: @shijinkui  thanks for the review...it is indeed redundant as i am catching actually all cases in the match pattern matching. I will remove it

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/table/plan/nodes/datastream/DataStreamJoin.scala [hunk 3]
```diff
@@ -0,0 +1,241 @@
...
+    if (first != null) {
+      val outrez = new Row(first.getArity + secondarity)
+      var i = 0
+      while (i < first.getArity) {
+        outrez.setField(i, first.getField(i))
+        i += 1
+      }
+      i = 0
+      while (i < secondarity) {
+        outrez.setField(i, secondR.getField(i))
```
Comment (rtudoran) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/table/plan/nodes/datastream/DataStreamJoin.scala]: field needs to account for the values from left side - i + first.getArity

**Comment by fhueske** [comment]
FLINK-9714 has been added and will be included in the next release. 
Closing this PR.

**Pull request closed without merging**